SimPEG/Directives.py

```diff
@@ -233,7 +250,12 @@ def target(self):
         if getattr(self, '_target', None) is None:
             # the factor of 0.5 is because we do phid = 0.5*|| dpred - dobs||^2
             if self.phi_d_star is None:
-                self.phi_d_star = 0.5 * self.survey.nD
+
+                nD = 0
+                for survey in self.survey:
+                    nD += survey.nD
```

Comment:
How we want to apply this is a bit subtle. I am not sure if we want to have a single stopping criteria for all phi_d's. If I am jointly inverting DC (nD = 50) and Richards (nD = 20), then we might want to set the stopping criteria such that we fit both the DC (reaching phi_d* = 25) and Richards (reaching phi_d*= 10) independently rather than reaching a "global" target misfit with total (phi_d* = 35).

Reply:
@lheagy You are right! My experiences of doing joint inversion is that, we need to fit each of the observed geophysical data set to about its noise level. Therefore, instead of calculating and monitoring the sum of all Phids, it is important to make sure that each phid actually reaches its target data misfit value. 

This problem become more critical when the number of data from one set of observations is very different from that from the other (say, nD = 200 for DC, nD = 20 for Richards), because the data set with more observations will be likely to dominate the minimization process (because decreasing the data misfit for the data set with more observations will lead to a quick decrease in the total objective function). 

Another way of looking at this problem is, consider the gradient, because we rely on gradient (or some sort of variant of gradient, like conjugate gradient) to minimize the objective function. It is very likely that the gradient will be dominated by the data set with more observations, as well.

You can do a quick test by setting nD for DC equal a very large number and nD for Richards equa a very small number, and letting the inversion only return a desired 'global' target.  A good chance you will see what I described above.

This is an open question for joint inversion of, say, seismic and gravity, as well.

I agree with you that, setting a target for a 'global' misfit is not a good idea.

Reply:
however, for the global / local problems, we will want a single `phi_d_star`. Maybe we use an option
`joint_inversion = True`, which does
```
self.phi_d_star = [0.5 * self.chifact * survey.nD for survey in self.survey]  
```

Reply:
You guys are the best @jiajiasun + @lheagy . I ll have another go at it.

@jiajiasun : Were your problems updating the same physical property (i.e. single vector of density) or one for each? If a single vector of density, do you think the number of data matter - or would the sensitivities (gradient) matter more?

Reply:
@fourndo My experiences are mostly with joint inversion of, say, seismic and gravity that deals with different physical properties. But what I described previously also applies to the problem of inverting for, say, a single vector of resistivity, from two different electrical data sets. I guess you guys are more interested in the latter?

Let us consider the total objective function phi(m) = w1 * phi_d1 + w2 * phi_d2 + beta * phi_m, where w1, w2 and beta are scalars, and phi_d1, phi_d2 and phi_m are data misfit and regularization terms. Imagine you set w1 = 1, and w2 = 0.00001, then the inversion would not bother trying to minimize phi_d2 because that term is already very very small due to the very very small value assigned to w2 when evaluating the total objective function. 

Similarly, suppose w1 = w2 = 1.0, but the number of data for the first set of data is 10000, and that for the second set of data is 10, then similar thing would happen; the inversion would focus much more on minimizing the first data misfit term, thereby allowing the first set of data to contribute more to the final inverted model.

Setting a 'global' target data misfit for the above two cases would most likely result in the second set of data contributing much less to the inversion.  Also, even if you achieve your 'global' target, you might end up with one set of data being overfit and the other being under-fit. Therefore, I would recommend monitoring both phid_d1 and phi_d2  and make sure they both achieve their respective target data misfit value.

So, when you have two sets of data with different amount of data, you might want to leverage w1 and w2 to make sure that both sets of data contribute approximately equal to the inversion process, assuming of course that you trust the two sets of data equals and you want them to both contribute to the final model.

We can also try to understand this problem by looking at gradient.  Suppose our total objective function is phi(m) = w1* ||d_1^obs - G1*m||_2^2 + w2* ||d_2^obs - G2*m||_2^2 + w3 * ||W_m*m||_2^2, then the gradient (i.e., our search direction) would be grad = w1*G_1^T*delta_d1 + w2*G_2^T*delta_d2  + beta*W_m^T*W_m*m. It is obvious that the first term is the contribution from the first set of data and so on. If w1 = w2 = 1.0, and suppose number of first set of data is much much larger than the number of second set of data, then it is very likely that G_1^T*delta_d1 will be much larger than G_2*T*delta_d2, because delta_d is data residual and G^T*delta_d is the projection of data residual onto data space spanned by the rows of the G matrix. (This is my guess. I haven't verified it numerically).

In this case, we might consider adjusting w1 and w2 to make sure that both sets of data contribute to the gradient on more or less the same level.

Again, this is still an open question for joint inversion (of multiple sets of data sensitive to different physical properties) because we often have very different number of seismic data than gravity data. I have not seen a  satisfactory answer and even an attempt to investigate this issue in literature. This is a challenging problem!  (What I wrote down above reflects my current thinking, which may change in a few month :)

Reply:
This is great Jiajia, 
I will set it up so everything becomes a list (phi_d*, beta's), just like we did for the misfit and regularization functions. That way we can monitor each individually. We interested in both versions (multiple misfits + multiple phys props  || multiple misfits + one phys props), which we should have the mechanics to tackle both very similarly, maybe just different flavors of directives.

Agreed, still a very open-question I think. Hopefully we can share experiments faster/easier once we have this guy set up properly.

Reply:
Would it just be a matter of dividing each data misfit term by the number of data? Or at least could that be a good starting point? Essentially it would mean each data misfit term contributes equally to the final objective model at the point when both of them are fit to their noise levels. However, I only have intuitions about how they contribute outside of that point.

Reply:
Hey @jcapriot ,
I like the way to think. I had try that before, and it does work in theory. In practice though, phi_d is also used in the optimization part also, and it becomes less transparent when we start messing up with the measure itself. Other option is to have multipliers on the misfit function (like a beta, but on phid). So that way we could change the weight of each regularizations AND misfit functions independantly. 
something like 
```
phi = beta1 * phid_1 + beta2 * phid_2 + beta3*reg1 + beta4*reg2 + ...
```
Default would be all 1
Thoughts?

Reply:
It is an option as well! Actually, some folks from UCSD used this trick in their joint inversion of, say, CSEM and MT, for resistivity. I forgot the names of the authors. Will find out later. 

My personal view on this option is that, if the noise from each set of data can be estimated well, then dividing by the number of data is a good idea. Otherwise, it might not be. But, I have not played with this strategy in my work yet. Need to run some inversions with all these different strategies and find out which one works best.

Reply:
@fourndo It might almost simplify the ``phi_d`` measure in the optimization, meaning they would only ever have to hit ``1.0`` instead of ``nD``.

@jiajiasun I feel like part of doing joint inversion should be estimating the noise ahead of time. I mean at some point there just gets to be too many unknowns at once, and it's not exactly feasible to black-box it.

Reply:
I would hesitate to divide by the number of data by default - I would suggest we leave it up to the user to define an appropriate multiplier (if we do something under the hood, it has the potential to be quite confusing). 

I could imagine scenarios where one data set has a very small number of data, another has a large number and you may want to effectively try treating all data equally (rather than treating each survey equally, as dividing by the number of data would do). I am sure each approach has its merits, but is should be the user who makes that decision and I think that keeping the definition of the data misfit quite simple should make things more transparent. 

my 2 cents :)

Reply:
Extending this to the joint inversion case will be a bit more involved. @fourndo: would you mind creating an issue and tagging it here? we can address this in a separate pr. For now, I think the current implementation is ok